File: api/streamchat.go
package handler

import (
	"bufio"
	"bytes"
	"encoding/json"
	"io"
	"net/http"
)

type Message struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

type ChatRequest struct {
	Model    string    `json:"model"`
	Messages []Message `json:"messages"`
	Stream   bool      `json:"stream"`
}

type Usage struct {
	PromptTokens     int `json:"prompt_tokens"`
	CompletionTokens int `json:"completion_tokens"`
	TotalTokens      int `json:"total_tokens"`
}

type Choice struct {
	Message      Message `json:"message"`
	FinishReason string  `json:"finish_reason"`
	Index        int     `json:"index"`
}

type ErrorResp struct {
	Type string `json:"type"`
}

type ChatResponse struct {
	ID      string    `json:"id"`
	Object  string    `json:"object"`
	Created int       `json:"created"`
	Model   string    `json:"model"`
	Usage   Usage     `json:"usage"`
	Choices []Choice  `json:"choices"`
	Error   ErrorResp `json:"error"`
}

type APIResponse struct {
	Success bool        `json:"success"`
	Result  interface{} `json:"result"`
	Message string      `json:"message"`
}

var httpClient = http.Client{}
var API_KEY = "Bearer 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"

func StreamChatHandler(w http.ResponseWriter, r *http.Request) {
	if r.Method == http.MethodPost {
		request := ChatRequest{}
		err := json.NewDecoder(r.Body).Decode(&request)
		if err != nil {
			http.Error(w, err.Error(), http.StatusBadRequest)
			return
		}
		messageLen := len(request.Messages)
		if messageLen > 10 {
			request.Messages = request.Messages[messageLen-10:]
		}
		request.Stream = true

		reqByte, err := json.Marshal(request)
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}

		req, err := http.NewRequest("POST", "https://api.openai.com/v1/chat/completions", bytes.NewReader(reqByte))
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		req.Header.Set("Content-Type", "application/json")
		req.Header.Set("Authorization", API_KEY)

		resp, err := httpClient.Do(req)
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		reader := bufio.NewReader(resp.Body)
		defer resp.Body.Close()

		_, err = io.Copy(w, reader)
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}

		return

	} else {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}

}

File: api/streamchat.py
from http.server import BaseHTTPRequestHandler
import os
import requests


class Handler(BaseHTTPRequestHandler):

    def do_POST(self):
        self.send_response(200)
        self.send_header('Content-type', 'application/json')
        self.end_headers()

        content_len = int(self.headers.get('content-length', 0))
        post_body = self.rfile.read(content_len)
        print(post_body.decode("utf-8"))

        # send to openai
        url = "https://api.openai.com/v1/chat/completions"
        headers = {
            "Content-Type": "application/json",
            'Authorization': 'Bearer ' + os.getenv("OPENAI_API_KEY")
        }

#         requests.post(url, stream=True, headers=headers, data=post_body)
#         self.wfile.write(b'{"hello":"world"}')

        with requests.post(url, stream=True, headers=headers, data=post_body) as response:
            for chunk in response.iter_content(chunk_size=512):
                self.wfile.write(chunk)
                break
